internal/batches/executor/run_steps.go

```diff
@@ -78,27 +78,44 @@ func runSteps(ctx context.Context, opts *executionOpts) (result executionResult,
 	}
 	defer workspace.Close(ctx)
 
-	execResult := executionResult{
-		Outputs: make(map[string]interface{}),
-		Path:    opts.task.Path,
-	}
-	results := make([]StepResult, len(opts.task.Steps))
+	var (
+		execResult = executionResult{
+			Diff:         "",
+			ChangedFiles: &git.Changes{},
+			Outputs:      make(map[string]interface{}),
+			Path:         opts.task.Path,
+		}
+		previousStepResult StepResult
+		startStep          int
+	)
 
-	var startStep int
 	if opts.task.CachedResultFound {
 		// Set the Outputs to the cached outputs
 		execResult.Outputs = opts.task.CachedResult.Outputs
 
 		startStep = opts.task.CachedResult.StepIndex + 1
 
+		// If we have cached results and don't need to execute any more steps,
+		// we can quit
+		if startStep == len(opts.task.Steps) {
+			changes, err := git.ChangesInDiff(opts.task.CachedResult.Diff)
+			if err != nil {
+				return execResult, nil, errors.Wrap(err, "parsing cached step diff")
+			}
+
+			execResult.Diff = string(opts.task.CachedResult.Diff)
+			execResult.ChangedFiles = &changes
+			stepResults = append(stepResults, opts.task.CachedResult)
+
+			return execResult, stepResults, nil
+		}
+
```

Comment:
This is the fix for the problem this PR aims to address. The rest of the changes are either tests or were made to make this fix easy to make (like the removal of the whole `results` indexing, introducing `previousStepResult`, etc.)